[Flang] Minloc elemental intrinsic lowering

davemgreen opened the pull request:
Currently the lowering of a minloc intrinsic with a mask will look something like
```
  %e = hlfir.elemental %shape ({
    ...
  })
  %m = hlfir.minloc %array mask %e
  hlfir.assign %m to %result
  hlfir.destroy %m
```
The elemental will be expanded into a temporary+loop, the minloc into a
FortranAMinloc call (which hopefully gets simplified to a specialized call that
can be inlined at the call site), and the assign might get expanded to a
FortranAAssign. The assign we could inline too, but it would be better to
generate the entire construct as single loop if we can - one that performs the
minloc calculation with the mask elemental computed inline and assigns directly
to the output array.

This patch attempt to do that, adding a hlfir version of the expansion code
from SimplifyIntrinsics that turns an assign+minloc+elemental into a single
combined loop nest. It attempts to reuse the methods in genMinlocReductionLoop
for constructing the loop with a modified loop body. The declaration for the
function is currently in Optimizer/Support/Utils.h, but there might be a better
place for it.

It is currently added as port of the OptimizedBufferizationPass. I originally
had it as part of the SimplifyHLFIRIntrinsics pass, but there were already some
methods doing similar things in OptimizedBufferization. It just needs to happen
before the elementals are expanded. I think I would like to do a similar thing
for maxloc and any/all/count too if this looks OK. I will rebase over #74436
once that goes in.

tblah commented on flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp:
```diff
@@ -659,6 +677,181 @@ mlir::LogicalResult VariableAssignBufferization::matchAndRewrite(
+
+      for (unsigned int i = 0; i < rank; ++i) {
+        mlir::Value index = builder.createIntegerConstant(loc, idxTy, i);
+        mlir::Value resultElemAddr = builder.create<fir::CoordinateOp>(
```
Please use hlfir::DesignateOp for indexing in hlfir code

davemgreen replied:
There is still one use of Coordicate left from the shared code. Do you think it is worth trying to change it over too?

tblah replied:
I'm not sure. I guess not. @jeanPerier do you have a strong feeling on this?

tblah replied:
I guess if CoordinateOp works, that is because we already bufferized the hlfir.expr. Which I guess makes sense in OptimizedBufferization, but not if you follow Jean's suggestion to move this to SimplifyHLFIRIntrinsics

tblah reviewed (commented):
This looks great! Really nice work. Just two comments

jeanPerier reviewed (commented):
Was there a lot of duplication rewriting it to an hlfir.elemental in in SimplifyHLFIRIntrinsics?

I would favor doing it there because I think having it as an hlfir.elemental is more powerful in general since it can then be composed if it is used as an operand to some other array operation.

Think: `array1 = offset + minloc(array2, dim=dim)`

That is the direction I wanted to go to implement https://github.com/llvm/llvm-project/blob/main/flang/docs/ArrayComposition.md and avoid array temporaries as much as possible.

vzakhari commented on flang/lib/Optimizer/Transforms/SimplifyIntrinsics.cpp:
```diff
@@ -523,27 +475,12 @@ genMinlocReductionLoop(fir::FirOpBuilder &builder, mlir::func::FuncOp &funcOp,
-    // mlir::Value one = builder.createIntegerConstant(loc, resultElemType, 1);
     mlir::Value index = builder.createIntegerConstant(loc, idxTy, i);
     mlir::Value resultElemAddr =
         builder.create<fir::CoordinateOp>(loc, resultRefTy, resultArr, index);
```
I would prefer this to be done via `DesignateOp` for the HLFIR path.

vzakhari commented on flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp:
```diff
@@ -659,6 +677,194 @@ mlir::LogicalResult VariableAssignBufferization::matchAndRewrite(
+    if (!destroy)
+      return rewriter.notifyMatchFailure(assign, "Did not find destroy");
+
+    if (!checkForElementalEffectsBetween(elemental, assign, minloc.getArray(),
```
Please make sure that the effects are handled properly in this case:
```
function test(x, mask)
  integer :: x(:)
  integer :: mask(:)
  print *, x(5:5)
  x(5:5) = minloc(x,mask=mask.ge.5)
end function test
```

Compiling this with -O2 gives this before the optimized bufferization:
```
    %3:2 = hlfir.declare %arg0 {uniq_name = "_QFtestEx"} : (!fir.box<!fir.array<?xi32>>) -> (!fir.box<!fir.array<?xi32>>, !fir.box<!fir.array<?xi32>>)
...
    %8 = hlfir.designate %3#0 (%c5:%c5:%c1)  shape %7 : (!fir.box<!fir.array<?xi32>>, index, index, index, !fir.shape<1>) -> !fir.box<!fir.array<1xi32>>
...
    %14 = hlfir.elemental %13 unordered : (!fir.shape<1>) -> !hlfir.expr<?x!fir.logical<4>> {
    ^bb0(%arg2: index):
      %17 = hlfir.designate %0#0 (%arg2)  : (!fir.box<!fir.array<?xi32>>, index) -> !fir.ref<i32>
      %18 = fir.load %17 : !fir.ref<i32>
      %19 = arith.cmpi sge, %18, %c5_i32 : i32
      %20 = fir.convert %19 : (i1) -> !fir.logical<4>
      hlfir.yield_element %20 : !fir.logical<4>
    }
    %15 = hlfir.minloc %3#0 mask %14 {fastmath = #arith.fastmath<contract>} : (!fir.box<!fir.array<?xi32>>, !hlfir.expr<?x!fir.logical<4>>) -> !hlfir.expr<1xi32>
    hlfir.assign %15 to %8 : !hlfir.expr<1xi32>, !fir.box<!fir.array<1xi32>>
```

davemgreen replied:
With the latest version of the code this is hopefully getting fixed in #77899.

psteinfeld reviewed (commented):
I've built and tested this patch successfully.  I ran both check-flang and NVIDIA's internal tests, and all looks good.

I don't understand the code very well, so someone who understands the code should approve.

davemgreen commented:
Thanks. That's good to hear. I'll try and get back to addressing the review comments soon, some of the comments here need more work to not produce worse performance. I was hoping to focus on #75774 first.

davemgreen commented on flang/lib/Optimizer/HLFIR/Transforms/OptimizedBufferization.cpp:
```diff
@@ -659,6 +677,194 @@ mlir::LogicalResult VariableAssignBufferization::matchAndRewrite(
+// inlined elemental and no extra temporaries.
+//  %e = hlfir.elemental %shape ({ ... })
+//  %m = hlfir.minloc %array mask %e
+//  hlfir.assign %m to %result
```
I've had a go at making this work as minloc(elemental), not handling the assign as the same time. A lot of the benefit of this patch was from aiming to produce less temporaries, writing straight into the result is going to produce better final code than having multiple temporaries being written and copied again and again.

But I think it can work OK here so long as the assign is simple enough and get optimized away in the end. The one thing it needs to be careful of is to make sure that any assign uses of the minloc is applied to the temp array, not a new asexpr, so that the VariableAssignBufferization from this pass has a chance to kick-in. Let me know what you think.

davemgreen commented:
Ping, thanks.

tblah reviewed (approved):
LGTM, thanks for this

davemgreen committed:
[Flang] Minloc elemental intrinsic lowering

Currently the lowering of a minloc intrinsic with a mask will look something
like:
  %e = hlfir.elemental %shape ({
    ...
  })
  %m = hlfir.minloc %array mask %e
  hlfir.assign %m to %result
  hlfir.destroy %m
The elemental will be expanded into a temporary+loop, the minloc into a
FortranAMinloc call (which hopefully gets simplified to a specialized call that
can be inlined at the call site), and the assign might get expanded to a
FortranAAssign. It would be better to generate the entire construct as single
loop if we can - one that performs the minloc calculation with the mask
elemental computed inline.

This patch attempt to do that, adding a hlfir version of the expansion code
from SimplifyIntrinsics that turns an minloc+elemental into a single combined
loop nest. It attempts to reuse the methods in genMinlocReductionLoop for
constructing the loop with a modified loop body. The declaration for the
function is curently in Optimizer/Support/Utils.h, but there might be a better
place for it.

It is currently added as port of the OptimizedBufferizationPass, like the
similar count/any/all that have been added recently.

davemgreen commented:
Thanks. I gave this another test and it looks like it did OK. I will push this now, let me know if any problems come up or anything needs adjusting.

I'll try and expand this to maxloc too, to keep them more symmetrical.
